### ISSUE TITLE
Unable to repro dreamer_v3_100k_ms_pacman experiment reward results

### ISSUE THREAD

**Issue by brodequin-slaps** [Eclectic-Sheep/sheeprl#228]
Hi,

Ran the default main branch on the dreamer_v3_100k_ms_pacman experiment (seed 5), but could not repro the rewards advertised 

Advertised curve:
<img width="674" alt="pacman100k_dreamer_v3" src="https://github.com/Eclectic-Sheep/sheeprl/assets/15318348/0ba01528-10ef-4774-ae3b-2e8794f9e493">

When I run it locally with defaulted everything:
![my_pacman100k_run](https://github.com/Eclectic-Sheep/sheeprl/assets/15318348/d8cc2382-d3aa-4d23-90ee-121870091a50)

Wondering what could explain the difference?

Edit: Found out about deterministic mode which is disabled by default. Will update with the deterministic run results once finished

Edit: Finished run:

![image](https://github.com/Eclectic-Sheep/sheeprl/assets/15318348/69fa0f99-584b-4fba-b5b3-80622911e66a)

**Comment by michele-milesi** [comment]
Hi @brodequin-slaps, thanks for reporting this.
In the meantime, I will check that there are no reproducibility problems in our code.

PS. We run that experiment with `cfg.torch_deterministic=False`

**Comment by belerico** [comment]
Could it be the seed? I've also seen a large variance in Hafner results' on some environments given different seeds. We can maybe try to run another experiment with a different seed?

**Comment by belerico** [comment]
Also, which SheepRL version or commit are you using?

**Comment by brodequin-slaps** [comment]
Hi @belerico ,

Using commit `e8a68f33dac5684c2dc0659c31ff8999d58659c5`

Since using the same seed as upstream, it would make sense to me that the results obtained would match those advertised (especially if ran with `cfg.torch_deterministic=True`), that way it can give potential library users early confidence that they can repro advertised results. Maybe once my deterministic pacman run finishes (it's about a 2x slowdown - should finish tomorrow), someone could try it too with the same seed to see if it matches

**Comment by michele-milesi** [comment]
Sure, I try to run other experiments.
In the meantime, can you share with us your torch version and the version of your cuda drivers?

Thanks

**Comment by brodequin-slaps** [comment]
```bash
(.venv) sam@sam:~/dev/ml/sheeprl$ python -c "import torch; print(torch.__version__)"
2.2.1+cu121 # Torch

(.venv) sam@sam:~/dev/ml/sheeprl$ nvidia-smi
Wed Mar  6 08:36:26 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.104.07             Driver Version: 537.34       CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce GTX 1080        On  | 00000000:02:00.0  On |                  N/A |
| 49%   64C    P0             152W / 200W |   7696MiB /  8192MiB |    100%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```

**Comment by belerico** [comment]
Hi @brodequin-slaps, can you try out [this branch](https://github.com/Eclectic-Sheep/sheeprl/tree/fix/determinism)?
You can try to run a reproducible experiment with the following settings:

```bash
python sheeprl.py \
exp=dreamer_v3_100k_ms_pacman \
fabric.devices=1 \
fabric.accelerator=cuda \
torch_use_deterministic_algorithms=True \
torch_backends_cudnn_benchmark=False \
torch_backends_cudnn_deterministic=True \
cublas_workspace_config=":4096:8"
```

where `cublas_workspace_config=":4096:8"` comes from [here](https://docs.nvidia.com/cuda/cublas/index.html#results-reproducibility), while `torch_use_deterministic_algorithms=True`, `torch_backends_cudnn_benchmark=False` and `torch_backends_cudnn_deterministic=True` comes from the [PyTorch reproducibility page](https://pytorch.org/docs/stable/notes/randomness.html).

While I haven't tried specifically with Dreamer-V3, I've some simple and fast expriments with PPO:

![image](https://github.com/Eclectic-Sheep/sheeprl/assets/18405289/bd516525-caa5-4a8a-a147-01675b909a4f)

where:

* the ${\color{fuchsia}\text{fuchsia}}$ and ${\color{lightskyblue}\text{lightblue}}$ are run **without the deterministic features**
* the ${\color{orange}\text{orange}}$ and ${\color{blue}\text{blue}}$ (which cannot be seen since the orange one perfectly overlaps with it) are run **with the deterministic features**

**Comment by brodequin-slaps** [comment]
Hi @belerico ,

Tried the `fix/determinism` branch, and locally, they runs are deterministic
![image](https://github.com/Eclectic-Sheep/sheeprl/assets/15318348/e9e4d7da-e55a-435e-9c9e-6982d94ed2d6)

However, they don't correspond to your experiments, maybe something's different in our setup

**Comment by belerico** [comment]
Hi @brodequin-slaps, I don't think that one can achieve the perfect determinism on completely different hardware: https://discuss.pytorch.org/t/how-to-get-determistic-behavior-with-different-gpus/125640
